[XLA:CPU][oneDNN] Enable oneDNN Layer-Norm Custom Calls in Thunk Runtime

othakkar opened the pull request:
This PR enables support for oneDNN Layer-Norm operations in the XLA:CPU Thunk runtime, building upon the foundational implementation of `OneDnnOpThunk`.

Key changes:
- Updated thunk emitter to emit `OneDnnOpThunk` for oneDNN layer-norm during compilation.
- Re-enable the custom call rewrite to oneDNN layer-norm in `onednn_ops_rewriter.cc`.
- Added support for oneDNN Layer-Norm op via `ExecuteOneDnnLayerNorm(...)` in `onednn_layer_norm.cc`.

othakkar committed:
enable oneDNN layer norm in thunk runtime

othakkar committed:
address review comments

penpornk reviewed (changes requested):
Thank you for the PR and sorry for the delay!

penpornk reviewed (changes requested):
Could you please also add a unit test in `onednn_op_thunk_test.cc`?

othakkar commented on xla/service/cpu/onednn_layer_norm.cc:
```diff
@@ -42,10 +40,49 @@ using dnnl::engine;
 using dnnl::stream;
 }  // namespace
 
+void ExecuteOneDnnLayerNorm(absl::Span<MemrefInfoHandler> arguments,
```
I plan to submit a cleanup PR for all four oneDNN primitives to remove all dead code related to the legacy runtime, later, once the thunk runtime changes are merged. This is to avoid making this (and other thunk) PRs longer than needed. Let me know if that is okay.

penpornk replied:
I think it's not too complicated to clean them up in the same PR. This will make it easier to read the diff because most contents in `ExecuteOneDnnLayerNorm` will be shown as existing code and don't need to be reviewed. We can wait and see how it goes with the convolution thunk PR, but even if that one looks a bit messy, the code in this PR is much shorter and should look pretty clean.

othakkar committed:
Merge branch 'main' into othakkar/onednn_thunk_layer_norm_nonFFI

othakkar committed:
adding a unit test for LN

penpornk commented on xla/service/cpu/onednn_ops_rewriter.cc:
```diff
@@ -596,6 +596,9 @@ class OneDnnOpsRewriterVisitor : public DfsHloRewriteVisitor {
+    // TODO(intel-tf): remove this restriction after adding oneDNN softmax
+    // support in thunk runtime.
+    return absl::OkStatus();
     if (divide_instr->HasControlDependencies()) return absl::OkStatus();
```
Please wrap the rest of the code in this function with 
```
// NOLINTBEGIN(clang-diagnostic-unreachable-code)
...
// NOLINTEND(clang-diagnostic-unreachable-code)
```
so the compiler won't complain about unreachable code.

othakkar replied:
Done (https://github.com/openxla/xla/pull/31707/commits/00ffbac879ca18ec6c411b94bfab4c3025168dec).

othakkar committed:
remove dead-code for layer-norm related to legacy runtime

othakkar committed:
add lint-related tags to onednn_ops_rewriter

othakkar committed:
replace auto with corresponding type

othakkar commented on xla/backends/cpu/runtime/onednn/onednn_op_thunk_test.cc:
```diff
@@ -106,6 +104,88 @@ TEST(OneDnnOpThunkTest, SimpleOneDnnMatMulThunk) {
+
+  // Compare with tolerance
+  auto exp_view = expected.data<float>();
+  auto out_view = out_literal.data<float>();
```
Done (https://github.com/openxla/xla/pull/31707/commits/e212f9b62d30139c19993bb2ac747628cbf95763).

othakkar commented:
@penpornk I added a commit (https://github.com/openxla/xla/pull/31707/commits/e329036b050a7a2dbec33800fb3b9807de2e864b) to remove the legacy runtime code for oneDNN LayerNorm. Let me know if this looks good. I also addressed your other comments. Thanks!

othakkar committed:
Merge remote-tracking branch 'upstream/main' into othakkar/onednn_thunk_layer_norm_nonFFI

othakkar committed:
minor fix in onednn/BUILD

penpornk reviewed (changes requested):
Could you please merge conflicts here as well?

othakkar committed:
Merge remote-tracking branch 'upstream/main' into othakkar/onednn_thunk_layer_norm_nonFFI

othakkar committed:
move the NOLINTBEGIN position in onednn_ops_rewriter

othakkar committed:
clang-format checks

penpornk reviewed (approved):
Thank you again for the PR!
